Add documentation (#117)
stacks-network/rendezvous: docs/chapter_6.md

```

Let's break down each part of the command.

### Positional Arguments

Consider this example Clarinet project structure:

```
root
├── Clarinet.toml
├── contracts
│   ├── contract.clar
│   ├── contract.tests.clar
└── settings
    └── Devnet.toml
```

**1. Path to the Clarinet Project**

The `<path-to-clarinet-project>` is the relative or absolute path to the root directory of the Clarinet project. This is where the `Clarinet.toml` file exists. **It is not the path to the `Clarinet.toml` file itself**.

For example, if you're in the parent directory of `root`, the correct relative path would be:

```bash
rv ./root <contract-name> <type>
```

**2. Contract Name**

The `<contract-name>` is the name of the contract to be tested, as defined in `Clarinet.toml`.

For example, if `Clarinet.toml` contains:

```toml
[contracts.contract]
path = "contracts/contract.clar"
```

To test the contract named `contract`, you would run:

```bash
rv ./root contract <type>
```

**3. Testing Type**

The `<type>` argument specifies the testing technique to use. The available options are:

- `test` – Runs property-based tests.
- `invariant` – Runs invariant tests.

> 

Reviewer: Here we can link/direct the readers to Chapter 4 which explains those testing types.

Author: Addressed in ad26b02.

**Running property-based tests**

To run property-based tests for the `contract` contract, ensure that your test functions are defined in:

```
./root/contracts/contract.tests.clar
```

Then, execute:

```bash
rv ./root contract test
```

This tells Rendezvous to:

- Load the **Clarinet project** located in `./root`.
- Target the **contract** named `contract` as defined in `Clarinet.toml` by executing **property-based tests** defined in `contract.tests.clar`.

**Running invariant tests**

> Invariant tests verify that certain properties of a contract remain consistent across multiple states. They are also defined in the test file:

Reviewer: ```suggestion
To run invariant tests for the contract contract, ensure that your test functions are defined in:
```

Reviewer: Since there is no explanation for property-based tests, it should be the same style also here. I've modified the text to be same as the one in property-based tests.

Author: Addressed in dc6b7da.

```
./root/contracts/contract.tests.clar
```

To run invariant tests, use:

```bash
rv ./root contract invariant
```

With this command, Rendezvous will:

- Randomly **execute public function calls** in the `contract` contract.

> - **Periodically check the defined invariants** to ensure the contract's internal state remains valid.

Reviewer: ```suggestion
- **Randomly check the defined invariants** to ensure the contract's internal state remains valid.
```

Author: I think `periodically` describes the Rendezvous invariant testing behavior more closely.

| `<path-to-clarinet-project>` | Path to the Clarinet project (where `Clarinet.toml` is located).                 | `rv root contract test`                           |
| `<contract-name>`            | Name of the contract to test (as in `Clarinet.toml`).                            | `rv root contract test`                           |
| `<type>`                     | Type of test (`test` for property-based tests, `invariant` for invariant tests). | `rv root contract test`                           |
| `--runs=<num>`               | Sets the number of test iterations (default: 100).                               | `rv root contract test --runs=500`                |
| `--seed=<num>`               | Uses a specific seed for reproducibility.                                        | `rv root contract test --seed=12345`              |
| `--dial=<file>`              | Loads JavaScript dialers from a file for pre/post-processing.                    | `rv root contract test --dial=./custom-dialer.js` |

---

## Understanding Rendezvous

> Rendezvous treats **property-based tests** and **invariants** as first-class citizens of the target contract. This follows the principle that _effective tests should be written in the same language as the system under test and run as part of it_.

Reviewer: ```suggestion
Rendezvous makes **property-based tests** and **invariant tests** first-class. Tests are written in the same language as the system under test. This helps developers master the contract language. It also pushes boundaries—programmers shape their thoughts first, then express them using the language's tools.
```

Reviewer: Something like this? Maybe mention "Clarity" instead of just "language".

Author: Addressed. 🚀

# Usage

This chapter explains how to use Rendezvous in different situations. By the end, you’ll know when and how to use its features effectively.

> 

Author: For sure we'll need a ToC here.

Reviewer: Done https://github.com/stacks-network/rendezvous/pull/117/commits/a2bc65b411384557fc9ec6901ecf5802d723ab9d.

Both have access to an object containing:

- `selectedFunction` – The function being executed.
- `functionCall` – The result of the function call (`undefined` for **pre-dialers**).
- `clarityValueArguments` – The generated Clarity values used as arguments.

**Example: Post-Dialer for SIP-010**

Below is a **post-dialer** that verifies SIP-010 compliance by ensuring that the `transfer` function emits a print event containing the `memo`.

```js
async function postTransferSip010PrintEvent(context) {
  const { selectedFunction, functionCall, clarityValueArguments } = context;

  // Ensure this check runs only for the "transfer" function.
  if (selectedFunction.name !== "transfer") return;

  const functionCallEvents = functionCall.events;
  const memoParameterIndex = 3; // The memo parameter is the fourth argument.

  const memoGeneratedArgumentCV = clarityValueArguments[memoParameterIndex];

  // If the memo argument is `none`, there's nothing to validate.
  if (memoGeneratedArgumentCV.type === 9) return;

  // Ensure the memo argument is an option (`some`).
  if (memoGeneratedArgumentCV.type !== 10) {
    throw new Error("The memo argument must be an option type!");
  }

> 

Reviewer 2: the comparison operation value is not update like in the https://github.com/stacks-network/rendezvous/blame/master/example/sip010.cjs
